File: 백준/15684_사다리조작/Main.java
class Solution {

    private int N, M, H;

    private boolean[][] visited;

    public Solution(BufferedReader reader) throws IOException {
        String[] tokens = reader.readLine().split(" ");
        N = Integer.parseInt(tokens[0]);
        M = Integer.parseInt(tokens[1]);
        H = Integer.parseInt(tokens[2]);

        visited = new boolean[N+2][H+1];

        for(int i = 0 ; i < M ; i++) {
            tokens = reader.readLine().split(" ");
            int a = Integer.parseInt(tokens[0]);
            int b = Integer.parseInt(tokens[1]);
            visited[b][a] = true;
        }
    }

    private boolean playGame() {
        boolean result = true;

        for(int i = 1 ; i <= N ; i++) {
            int cur = i;

            for(int j = 1 ; j <= H ; j++) {
                if(visited[cur][j]) {
                    cur++;
                } else if(visited[cur-1][j]) {
                    cur--;
                }
            }

            if(cur != i) {
                return false;
            }
        }

        return result;
    }

    private boolean combination(int depth, int from, int limit) {
        if(depth == limit) {
             return playGame();
        }

        boolean result = false;
        for(int i = from ; i < N ; i++) {
            for(int j = 1 ; j <= H ; j++) {
                if(visited[i][j] || visited[i-1][j] || visited[i+1][j]) {
                    continue;
                }

                visited[i][j] = true;
                result |= combination(depth + 1, i, limit);
                visited[i][j] = false;
                if(result) {
                    break;
                }
            }
        }

        return result;
    }

    public void run() {
        StringBuilder sb = new StringBuilder();
        int minValue = Integer.MAX_VALUE;

        for(int tryCount = 0 ; tryCount <= 3 ; tryCount++) {
            if(combination(0, 1, tryCount)) {
                minValue = tryCount;
                break;
            }
        }


        sb.append(minValue == Integer.MAX_VALUE ? -1 : minValue);
        System.out.println(sb.toString());
    }
}

File: 백준/15684_사다리조작/main.rs
trait Solution {
    fn new()->Self ;
    fn run(&mut self);
}

struct Problem {
    n: usize,
    m: usize,
    h: usize,
    visited: Vec<Vec<bool>>,
}

impl Solution for Problem {

    fn new() -> Self {
        let tokens: Vec<usize> = IO::read_line()
            .split_whitespace()
            .map(|x| x.parse().unwrap())
            .collect();

        let n = tokens[0];
        let m = tokens[1];
        let h = tokens[2];

        let mut visited = vec![vec![false; h+2]; n+2];

        for i in 0..m {
            let tokens: Vec<usize> = IO::read_line()
                .split_whitespace()
                .map(|x| x.parse().unwrap())
                .collect();
            visited[tokens[1]][tokens[0]] = true;
        }

        Self {
            n: n,
            m: m,
            h: h,
            visited: visited
        }
    }

    fn run(&mut self) {
        let mut minTryCount = i32::MAX;
        for tryCount in 0..4 {
            if self.dfs(0, 1, tryCount) {
                minTryCount = tryCount;
                break;
            }
        }

        if minTryCount == i32::MAX {
            minTryCount = -1;
        }

        println!("{}", minTryCount);
    }
}

impl Problem {

    fn check(&self) -> bool {
        let mut result = true;

        for i in 1..self.n+1 {
            let mut cur = i;

            for r in 1..self.h+1 {
                if self.visited[cur][r] {
                    cur += 1;
                } else if self.visited[cur-1][r] {
                    cur -= 1;
                }
            }

            if cur != i {
                result = false;
                break;
            }
        }

        result
    }

    fn dfs(&mut self, depth: i32, col: usize, limit: i32) -> bool {
        if depth == limit {
            return self.check()
        }

        for c in col..self.n {
            for r in 1..self.h+1 {

                if self.visited[c][r] || self.visited[c-1][r] || self.visited[c+1][r] {
                    continue;
                }

                self.visited[c][r] = true;
                let res = self.dfs(depth+1, c, limit);
                self.visited[c][r] = false;

                if(res) {
                    return res;
                }
            }
        }

        false
    }
}

fn main() {
    let mut problem: Problem = Problem::new();
    problem.run();
}

mod IO {
    pub fn read_line() -> String {
        let mut buf = String::new();
        std::io::stdin().read_line(&mut buf).unwrap();
        buf.trim().to_string()
    }

    pub fn read_copy_line(buf: &mut String) {
        std::io::stdin().read_line(buf).unwrap();
    }
}
